### ISSUE TITLE
WPS221 treats multi-line strings as a single line

### ISSUE THREAD

**Issue by gorrog** [wemake-services/wemake-python-styleguide#1758]
# Bug report

## What's wrong

The following code triggers a WPS221 error on line 2.
```
        return (
            f"Organisation code: {self.organisation_code}\n"
            f"Inventory code: {self.inventory_code}\n"
            f"Date string: {self.date_string}\n"
            f"Random string: {self.random_string}"
        )
```

## How it should be

Each string section should be treated as a separate line. While the whole string may be complex, each line of it is rather simple, so should not trigger a complexity warning.

## System information

Ubuntu 20.4

### flake8 information

Contents of `flake8 --bug-report`:

```
{
  "dependencies": [],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.8.6",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "flake8-bandit",
      "version": "2.1.2"
    },
    {
      "is_local": false,
      "plugin": "flake8-broken-line",
      "version": "0.2.1"
    },
    {
      "is_local": false,
      "plugin": "flake8-bugbear",
      "version": "19.8.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-comprehensions",
      "version": "3.3.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-darglint",
      "version": "1.5.8"
    },
    {
      "is_local": false,
      "plugin": "flake8-debugger",
      "version": "3.2.1"
    },
    {
      "is_local": false,
      "plugin": "flake8-django",
      "version": "0.0.4"
    },
    {
      "is_local": false,
      "plugin": "flake8-docstrings",
      "version": "1.5.0, pydocstyle: 5.1.1"
    },
    {
      "is_local": false,
      "plugin": "flake8-eradicate",
      "version": "0.3.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-print",
      "version": "3.1.4"
    },
    {
      "is_local": false,
      "plugin": "flake8-string-format",
      "version": "0.2.3"
    },
    {
      "is_local": false,
      "plugin": "flake8_commas",
      "version": "2.0.0"
    },
    {
      "is_local": false,
      "plugin": "flake8_isort",
      "version": "3.0.1"
    },
    {
      "is_local": false,
      "plugin": "flake8_multiline_containers",
      "version": "0.0.17"
    },
    {
      "is_local": false,
      "plugin": "flake8_quotes",
      "version": "2.1.2"
    },
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "naming",
      "version": "0.9.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.6.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.2.0"
    },
    {
      "is_local": false,
      "plugin": "rst-docstrings",
      "version": "0.0.12"
    },
    {
      "is_local": false,
      "plugin": "use-fstring-format",
      "version": "1.0"
    },
    {
      "is_local": false,
      "plugin": "use-fstring-percent",
      "version": "1.0"
    },
    {
      "is_local": false,
      "plugin": "use-fstring-prefix",
      "version": "1.0"
    },
    {
      "is_local": false,
      "plugin": "wemake_python_styleguide",
      "version": "0.14.1"
    }
  ],
  "version": "3.8.4"
}

```

### pip information

Contents of `pip freeze`:
```
alabaster==0.7.12
amqp==2.6.1
appdirs==1.4.4
argon2-cffi==20.1.0
asgiref==3.3.1
astor==0.8.1
astroid==2.4.2
attrs==20.3.0
autopep8==1.5.4
Babel==2.9.0
backcall==0.2.0
bandit==1.6.2
billiard==3.6.3.0
black==19.3b0
celery==4.4.6
certifi==2020.11.8
cffi==1.14.4
cfgv==3.2.0
chardet==3.0.4
click==7.1.2
coverage==5.3
cryptography==3.2.1
darglint==1.5.8
decorator==4.4.2
defusedxml==0.6.0
distlib==0.3.1
Django==3.0.11
django-allauth==0.43.0
django-celery-beat==2.1.0
django-cors-headers==3.5.0
django-coverage-plugin==1.8.0
django-crispy-forms==1.10.0
django-debug-toolbar==3.1.1
django-environ==0.4.5
django-extensions==3.0.9
django-model-utils==4.0.0
django-redis==4.12.1
django-stubs==1.7.0
django-timezone-field==4.1.1
djangorestframework==3.12.2
docutils==0.16
eradicate==1.0
factory-boy==3.1.0
Faker==4.17.1
filelock==3.0.12
flake8==3.8.4
flake8-bandit==2.1.2
flake8-broken-line==0.2.1
flake8-bugbear==19.8.0
flake8-commas==2.0.0
flake8-comprehensions==3.3.0
flake8-debugger==3.2.1
flake8-django==1.1.1
flake8-docstrings==1.5.0
flake8-eradicate==0.3.0
flake8-isort==3.0.1
flake8-multiline-containers==0.0.17
flake8-polyfill==1.0.2
flake8-print==3.1.4
flake8-quotes==2.1.2
flake8-rst-docstrings==0.0.12
flake8-string-format==0.2.3
flake8-use-fstring==1.1
flower==0.9.5
future==0.18.2
gitdb==4.0.5
GitPython==3.1.11
hiredis==1.1.0
humanize==3.1.0
identify==1.5.10
idna==2.10
imagesize==1.2.0
iniconfig==1.1.1
ipdb==0.13.4
ipython==7.19.0
ipython-genutils==0.2.0
isort==4.3.21
jedi==0.17.2
Jinja2==2.11.2
kombu==4.6.11
lazy-object-proxy==1.4.3
livereload==2.6.3
MarkupSafe==1.1.1
mccabe==0.6.1
mypy==0.790
mypy-extensions==0.4.3
nodeenv==1.5.0
oauthlib==3.1.0
packaging==20.7
parso==0.7.1
pbr==5.5.1
pep8-naming==0.9.1
pexpect==4.8.0
pickleshare==0.7.5
Pillow==8.0.1
pluggy==0.13.1
pre-commit==2.8.2
prometheus-client==0.8.0
prompt-toolkit==3.0.8
psycopg2-binary==2.8.6
ptyprocess==0.6.0
py==1.9.0
pycodestyle==2.6.0
pycparser==2.20
pydocstyle==5.1.1
pyflakes==2.2.0
Pygments==2.7.2
PyJWT==1.7.1
pylint==2.6.0
pyls-black==0.4.6
pyparsing==2.4.7
pytest==6.1.2
pytest-django==4.1.0
pytest-sugar==0.9.4
python-crontab==2.5.1
python-dateutil==2.8.1
python-jsonrpc-server==0.4.0
python-language-server==0.36.1
python-slugify==4.0.1
python3-openid==3.2.0
pytz==2020.4
PyYAML==5.3.1
redis==3.5.3
requests==1.0.0
requests-oauthlib==1.3.0
restructuredtext-lint==1.3.2
rope==0.18.0
six==1.15.0
smmap==3.0.4
snowballstemmer==2.0.0
Sphinx==3.3.1
sphinx-autobuild==2020.9.1
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==1.0.3
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.4
sqlparse==0.4.1
stevedore==3.2.2
termcolor==1.1.0
testfixtures==6.15.0
text-unidecode==1.3
toml==0.10.2
tornado==6.1
traitlets==5.0.5
typed-ast==1.4.1
typing-extensions==3.7.4.3
ujson==4.0.1
urllib3==1.26.2
vine==1.3.0
virtualenv==20.2.1
wcwidth==0.2.5
wemake-python-styleguide==0.14.1
Werkzeug==1.0.1
wrapt==1.12.1
yapf==0.30.0

```

**Comment by shaneaung** [comment]
Hi, I would like to claim this issue and work on it with a few peers.

**Comment by portablesounds** [comment]
Hi, I'd like to preface this by saying I'm working on this issue with @shaneaung .
One thing I'd like to point out is that WPS already has a rule against using implicit string concatenation (`WPS326`). I guess we should still consider the case where someone disables this rule, as implicit string concatenation is technically faster since it's computed at "compile time" (right before you run it) rather than run time.
This brings me to my second issue: the reason for this bug appears to be that the line is combined at compile time before the `AST` (which `WPS221` is based on) gets its hands on the code, so it will only be given one long string with the close quotes-newline-open quotes between each string completely gone. I don't know if there's a coherent way of checking this beyond directly analyzing the raw text in the file. Is this how `Tokenize` functions? Even if it does, combining tokenization to cross reference with the `AST` seems a bit heavy handed. Any thoughts?

**Comment by sobolevn** [comment]
Well, this is complicated. 

I don't consider this code valid at all:

```python
return (
            f"Organisation code: {self.organisation_code}\n"
            f"Inventory code: {self.inventory_code}\n"
            f"Date string: {self.date_string}\n"
            f"Random string: {self.random_string}"
        )
```

Because:
- It uses `f` strings, which are not allowed by `WPS`
- It uses implicit string concat which is not allowed by `WPS`

I would refactor this as:

```python
_SOME_TEMPLATE: Final = """
Organisation code: {organisation_code}
Inventory code: {inventory_code}
"""

# ...

returns _SOME_TEMPLATE.format(
    organisation_code=self.organisation_code,
)
```

**Comment by gorrog** [comment]
Thanks for the reply @sobolevn - I do see that this could be simplified to get around the violation, although I think there are times when you don't wish to do a 2 step template & format just to output a string when formatting in place is simpler to follow and results in less code. That said, I can see that there could be different opinions on this issue and since wemake style guide is the 'most opinionated linter ever' (paraphrasing), I don't expect you to take this any further.

Thank you for creating and maintaining this project - it has greatly improved the quality of my code!
